warehouse/router.go

```diff
@@ -0,0 +1,705 @@
+
+func (r *router) decrementActiveWorkers() {
+	r.activeWorkerCount.Add(-1)
+}
+
+func (r *router) getActiveWorkerCount() int {
+	return int(r.activeWorkerCount.Load())
+}
+
+func (r *router) setDestInProgress(warehouse model.Warehouse, jobID int64) {
+	r.inProgressMapLock.Lock()
+	defer r.inProgressMapLock.Unlock()
+
+	identifier := r.workerIdentifier(warehouse)
+	r.inProgressMap[WorkerIdentifierT(identifier)] = append(r.inProgressMap[WorkerIdentifierT(identifier)], JobID(jobID))
+}
+
+func (r *router) removeDestInProgress(warehouse model.Warehouse, jobID int64) {
+	r.inProgressMapLock.Lock()
+	defer r.inProgressMapLock.Unlock()
+
+	identifier := r.workerIdentifier(warehouse)
+
+	if idx, inProgress := r.checkInProgressMap(jobID, identifier); inProgress {
+		r.inProgressMap[WorkerIdentifierT(identifier)] = append(r.inProgressMap[WorkerIdentifierT(identifier)][:idx], r.inProgressMap[WorkerIdentifierT(identifier)][idx+1:]...)
+	}
+}
+
+func (r *router) isUploadJobInProgress(warehouse model.Warehouse, jobID int64) (int, bool) {
+	r.inProgressMapLock.RLock()
+	defer r.inProgressMapLock.RUnlock()
+
+	return r.checkInProgressMap(jobID, r.workerIdentifier(warehouse))
+}
+
+func (r *router) getInProgressNamespaces() []string {
+	r.inProgressMapLock.RLock()
+	defer r.inProgressMapLock.RUnlock()
+
+	var identifiers []string
+	for k, v := range r.inProgressMap {
+		if len(v) >= r.config.maxConcurrentUploadJobs {
+			identifiers = append(identifiers, string(k))
+		}
+	}
+	return identifiers
+}
+
+func (r *router) checkInProgressMap(jobID int64, identifier string) (int, bool) {
+	for idx, id := range r.inProgressMap[WorkerIdentifierT(identifier)] {
+		if jobID == int64(id) {
+			return idx, true
+		}
+	}
+	return 0, false
+}
+
+func (r *router) runUploadJobAllocator(ctx context.Context) {
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break loop
+		case <-r.bcManager.initialConfigFetched:
+			r.logger.Debugf("Initial config fetched in runUploadJobAllocator for %s", r.destType)
+		}
+
+		availableWorkers := r.config.noOfWorkers - r.getActiveWorkerCount()
+		if availableWorkers < 1 {
+			select {
+			case <-ctx.Done():
+				break loop
+			case <-time.After(r.config.waitForWorkerSleep):
+			}
+			continue
+		}
+
+		inProgressNamespaces := r.getInProgressNamespaces()
+		r.logger.Debugf(`Current inProgress namespace identifiers for %s: %v`, r.destType, inProgressNamespaces)
+
+		uploadJobsToProcess, err := r.uploadsToProcess(ctx, availableWorkers, inProgressNamespaces)
+		if err != nil {
+			if errors.Is(err, context.Canceled) ||
+				errors.Is(err, context.DeadlineExceeded) ||
+				strings.Contains(err.Error(), "pq: canceling statement due to user request") {
+
+				break loop
+			} else {
+				r.logger.Errorf(`Error executing uploadsToProcess: %v`, err)
+
+				panic(err)
+			}
+		}
+
+		for _, uploadJob := range uploadJobsToProcess {
+			r.setDestInProgress(uploadJob.warehouse, uploadJob.upload.ID)
+		}
+
+		for _, uploadJob := range uploadJobsToProcess {
```

Comment:
Is the double loop needed here? Could we potentially set the destination as "in progress" and give the job to a worker in the same iteration?